manta-crypto/src/signature.rs

```diff
@@ -0,0 +1,414 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Signature Schemes
+//!
+//! A signature scheme is made up of these three `trait`s:
+//!
+//! - [`Derive`]
+//! - [`Sign`]
+//! - [`Verify`]
+//!
+//! with the following completeness property:
+//!
+//! For all possible inputs, the following function returns `true`:
+//!
+//! ```text
+//! fn is_valid(signing_key: SigningKey, randomness: Randomness, message: Message) -> bool {
+//!     verify(derive(signing_key), message, sign(randomness, signing_key, message))
+//! }
+//! ```
+//!
+//! See the [`correctness`](test::correctness) test for more.
+
+/// Signing Key
+pub trait SigningKeyType {
+    /// Signing Key Type
+    type SigningKey;
+}
+
+impl<T> SigningKeyType for &T
+where
+    T: SigningKeyType,
+{
+    type SigningKey = T::SigningKey;
+}
+
+/// Verifying Key
+pub trait VerifyingKeyType {
+    /// Verifying Key Type
+    type VerifyingKey;
+}
+
+impl<T> VerifyingKeyType for &T
+where
+    T: VerifyingKeyType,
+{
+    type VerifyingKey = T::VerifyingKey;
+}
+
+/// Message
+pub trait MessageType {
+    /// Message Type
+    type Message;
+}
+
+impl<T> MessageType for &T
+where
+    T: MessageType,
+{
+    type Message = T::Message;
+}
+
+/// Signature
+pub trait SignatureType {
+    /// Signature Type
+    type Signature;
+}
+
+impl<T> SignatureType for &T
+where
+    T: SignatureType,
+{
+    type Signature = T::Signature;
+}
+
+/// Randomness
+pub trait RandomnessType {
+    /// Randomness Type
+    type Randomness;
+}
+
+impl<T> RandomnessType for &T
+where
+    T: RandomnessType,
+{
+    type Randomness = T::Randomness;
+}
+
+/// Signature Verifying Key Derivation Function
+pub trait Derive<COM = ()>: SigningKeyType + VerifyingKeyType {
+    /// Derives the verifying key from `signing_key`.
+    ///
+    /// This function is used by the signer to generate their [`VerifyingKey`] that is sent to the
+    /// verifier to check that the signature was valid.
+    ///
+    /// [`VerifyingKey`]: VerifyingKeyType::VerifyingKey
```

Comment:
Do we really need to document the return type here? I don't think it's necessary tho. Other methods in this file also don't have any return type documented so we'd better remove this for consistency?

Reply:
Also `Verify::Verification` is documented, so there's some precedent in this fil.e